compositor_pipeline/src/audio_mixer/internal_audio_mixer.rs

```diff
@@ -109,45 +126,159 @@ impl InternalAudioMixer {
+            sample
+        });
+
+        for (batch_index, input_samples) in samples_iter.enumerate() {
+            let sample_count = samples_in_frame.len();
+            let expected_next_sample_start_pts =
+                start_pts + Duration::from_secs_f64(sample_count as f64 / sample_rate as f64);
+
+            // potentially fill missing spots
+            if expected_next_sample_start_pts + max_error < input_samples.start_pts {
+                let missing_time = input_samples
+                    .start_pts
+                    .saturating_sub(expected_next_sample_start_pts);
+                let missing_samples_count = time_to_sample_count(missing_time);
+                if missing_samples_count < 1 {
+                    warn!(
+                        ?missing_time,
+                        "Distance between samples is higher than expected."
+                    )
+                }
+                samples_in_frame.extend((0..missing_samples_count).map(|_| (0i16, 0i16)))
+            }
+
+            let sample_count = samples_in_frame.len();
+            let expected_next_sample_start_pts =
+                start_pts + Duration::from_secs_f64(sample_count as f64 / sample_rate as f64);
+
+            // check if we need to drop samples at the beginning
+            let mut start_range = 0;
+            if expected_next_sample_start_pts > input_samples.start_pts + max_error {
+                let time_to_remove_from_start =
+                    expected_next_sample_start_pts.saturating_sub(input_samples.start_pts);
+                let samples_to_remove_from_start = time_to_sample_count(time_to_remove_from_start);
+                if batch_index != 0 {
+                    // We should only drop samples in the first batch.
+                    warn!(
+                        "Decoder produced overlapping samples. Dropping {} samples.",
+                        samples_to_remove_from_start
+                    );
+                }
+                start_range = samples_to_remove_from_start;
+            };
+
+            // check if we need to drop samples at the end
+            let mut end_range = input_samples.len();
+            if input_samples.end_pts > end_pts + max_error {
+                let desired_duration = end_pts.saturating_sub(expected_next_sample_start_pts);
+                let desired_sample_count = time_to_sample_count(desired_duration);
+                end_range = start_range + desired_sample_count;
+            }
+
+            samples_in_frame.extend(input_samples.samples[start_range..end_range].iter());
+        }
+
+        // Fill at the end only if last batch is ending to quickly
+        if let Some(pts) = last_batch_end_pts {
+            if pts < end_pts + max_error {
+                Self::ensure_correct_amount_of_samples(
+                    start_pts,
+                    end_pts,
+                    sample_rate,
+                    &mut samples_in_frame,
+                );
```

Comment:
You call the function that ensures the size and can potentially drain some samples. The if statement above should guarantee that it's not the case, but in similar places, you check for this explicitly and log errors. I don't like logging errors like this, but if you do it, make it consistent.